compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala

```diff
@@ -356,11 +369,11 @@ class ReifyQuotes extends MacroTransform {
             capturers(body.symbol)(body)
           case tree: RefTree if isCaptured(tree.symbol, level) =>
             val body = capturers(tree.symbol).apply(tree)
-            val splice: Tree =
-              if (tree.isType) body.select(tpnme.splice)
-              else ref(defn.InternalQuoted_exprSplice).appliedToType(tree.tpe).appliedTo(body)
-
-            transformSplice(body, splice)
+            if (tree.isType)
+              transformSpliceType(body, body.select(tpnme.spliceType))
+            else
+              val splice = ref(defn.InternalQuoted_exprSplice).appliedToType(tree.tpe).appliedTo(body).asInstanceOf[Apply]
```

Comment:
You could remove this cast by changing the return type of `def appliedTo`: https://github.com/lampepfl/dotty/blob/aa64078d9defe1df6856158a16c1f236792231f2/compiler/src/dotty/tools/dotc/ast/tpd.scala#L894-L899